runtime-sdk/src/error.rs

```diff
@@ -29,26 +29,171 @@ use crate::types::transaction::CallResult;
 /// ```
 pub trait Error: std::error::Error {
     /// Name of the module that emitted the error.
-    fn module_name() -> &'static str;
+    fn module_name(&self) -> &'static str;
 
     /// Error code uniquely identifying the error.
     fn code(&self) -> u32;
 
     /// Converts the error into a call result.
     fn to_call_result(&self) -> CallResult {
-        CallResult::Failed {
-            module: Self::module_name().to_owned(),
-            code: self.code(),
+        // If there is an underlying error, use that one.
+        match Error::source(self) {
+            Some(source) => source.to_call_result(),
+            None => CallResult::Failed {
+                module: self.module_name().to_owned(),
+                code: self.code(),
+            },
         }
     }
+
+    /// The lower-level source of this error, if any.
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        None
+    }
 }
 
 impl Error for std::convert::Infallible {
-    fn module_name() -> &'static str {
+    fn module_name(&self) -> &'static str {
         "(none)"
     }
 
     fn code(&self) -> u32 {
         Default::default()
     }
 }
+
+#[doc(hidden)]
+pub mod private {
+    use super::Error;
+
+    pub trait AsDynError {
+        fn as_dyn_error(&self) -> &(dyn Error + 'static);
+    }
+
+    impl<T: Error + 'static> AsDynError for T {
+        #[inline]
+        fn as_dyn_error(&self) -> &(dyn Error + 'static) {
+            self
+        }
+    }
+
+    impl AsDynError for dyn Error + 'static {
+        #[inline]
+        fn as_dyn_error(&self) -> &(dyn Error + 'static) {
+            self
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    const MODULE_NAME_1: &str = "test1";
+    const MODULE_NAME_2: &str = "test2";
+    const MODULE_NAME_3: &str = "test3";
+
+    #[derive(thiserror::Error, Debug, oasis_runtime_sdk_macros::Error)]
+    #[sdk_error(module_name = "MODULE_NAME_1")]
+    enum ChildError {
+        #[error("first error")]
+        #[sdk_error(code = 1)]
+        Error1,
+
+        #[error("second error")]
+        #[sdk_error(code = 2)]
+        Error2,
+    }
+
+    #[derive(thiserror::Error, Debug, oasis_runtime_sdk_macros::Error)]
+    #[sdk_error(module_name = "MODULE_NAME_2")]
+    enum ParentError {
+        #[error("first error")]
+        #[sdk_error(code = 1)]
+        NotForwarded(#[source] ChildError),
+
+        #[error("nested error")]
+        #[sdk_error(code = 2)]
+        Nested(
+            #[source]
+            #[sdk_source]
+            ChildError,
+        ),
+    }
+
+    #[derive(thiserror::Error, Debug, oasis_runtime_sdk_macros::Error)]
+    #[sdk_error(module_name = "MODULE_NAME_3")]
+    enum ParentParentError {
+        #[error("nested nested error")]
+        #[sdk_error(code = 1)]
+        Nested(
+            #[source]
+            #[sdk_source]
```

Comment:
So this actually means something different from _source_. It's actually marking the error as _transparent_. The error code isn't actually needed when setting the source either but is unfortunately required by a sane implementation of `code()`.

How about something like:
```
enum Error {
    #[sdk_error(code = 1, forward_to = ".0")
    Nested(#[source] ParentError)
}
```
where `foward_to` is a field of the variant

Reply:
Yeah this makes sense, it's just that the syntax seems uglier than annotating the field.

Reply:
Actually, it's not really forwarding everything since you can still get the whole chain by following `source()`. It's just that the default implementation of `to_call_result` will try to find the inner-most error.

Reply:
But maybe we can make it forward everything :-)

What if we call it `#[sdk_error(transparent)]` (without the code) and then it will forward (actually forward module, code and everything) the field that has the `#[source]` or `#[from]` attributes?

Reply:
I like it. I didn't want to suggest your writing more annotations code, though. I wonder whether thiserror shouldn't have collapsed it onto the field attr, though.

Reply:
transparent is a good convenience which I think will be good in some cases like gas accounting. but I'm worried that it won't be general enough. would there be cases where we do want context from the outer error?

if there were, for example a method that did two transfers atomically, we might desire to have errors like `two_transfers_module::Error::SecondTransferFail(source=accounts::Error::InsufficientBalance)`

Reply:
Added the `#[sdk_error(transparent)]` option, please take another look.

> if there were, for example a method that did two transfers atomically, we might desire to have errors like

Yeah but I don't see how we could fit this into a single module/code?